powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/OperationServiceBehavior.java

```diff
@@ -716,6 +734,102 @@ private Map<String, String> mapMerge(Map<String, String> m1, Map<String, String>
+            return null;
+        }
+
+        try {
+            byte[] seedBytes = Base64.getDecoder().decode(seed);
+            final LocalDateTime now = LocalDateTime.now();
+            byte[] totp = Totp.generateTotpSha256(seedBytes, now, otpLength);
+
+            final AuditDetail auditDetail = createProximityOtpAuditDetail(operation, seed, now);
+            audit.log(AuditLevel.INFO, "Proximity OTP generated for operation ID: {}", auditDetail, operationId);
+
+            return new String(totp);
+        } catch (CryptoProviderException | IllegalArgumentException e) {
+            logger.error("Unable to generate OTP for operation ID: {}, user ID: {}", operationId, operation.getUserId(), e);
+            throw new GenericServiceException(ServiceError.OPERATION_ERROR, e.getMessage(), e.getLocalizedMessage());
+        }
+    }
+
+    private static String generateTotpSeed(final OperationCreateRequest request, final OperationTemplateEntity template) throws GenericServiceException {
+        if (Boolean.FALSE.equals(request.getProximityCheckEnabled())) {
+            logger.debug("Proximity check is disabled in request from user ID: {}", request.getUserId());
+            return null;
+        } else if (Boolean.TRUE.equals(request.getProximityCheckEnabled()) || template.isProximityCheckEnabled()) {
+            logger.debug("Proximity check is enabled, generating TOTP seed for user ID: {}, templateName: {}", request.getUserId(), template.getTemplateName());
+            final KeyGenerator keyGenerator = new KeyGenerator();
+            try {
+                final byte[] seed = keyGenerator.generateRandomBytes(PROXIMITY_OTP_SEED_LENGTH);
+                return Base64.getEncoder().encodeToString(seed);
+            } catch (CryptoProviderException e) {
+                logger.error("Unable to generate proximity OTP seed for operation, user ID: {}", request.getUserId(), e);
+                throw new GenericServiceException(ServiceError.OPERATION_ERROR, e.getMessage(), e.getLocalizedMessage());
+            }
+        }
+        logger.debug("Proximity check not enabled neither in request user ID: {} nor in templateName: {}", request.getUserId(), template.getTemplateName());
+        return null;
+    }
+
+    private boolean proximityCheckPassed(final OperationEntity operation, final OperationApproveRequest request) {
+        final String seed = operation.getTotpSeed();
+        if (seed == null) {
+            return true;
+        }
+
+        final String otp = request.getAdditionalData().get(PROXIMITY_OTP);
+        if (otp == null) {
+            logger.warn("Proximity check enabled for operation ID: {} but proximity OTP not sent", operation.getId());
+            return false;
+        }
+        try {
+            final LocalDateTime now = LocalDateTime.now();
+            boolean result = Totp.validateTotpSha256(otp.getBytes(), Base64.getDecoder().decode(seed), now);
```

Comment:
I am missing the expected length in this method, it probably will just result in one if-else check.

Reply:
Validation of the length added.